File: Sources/logseq-compiler/link-finder.swift
//
//  File.swift
//  
//
//  Created by Deniz Aydemir on 6/4/22.
//

import Foundation


struct BlockPropertyFinder {
    private func pattern() -> String {
        return #"\n\S+::\s+\S+"#
    }
    
    private func regex() -> NSRegularExpression? {
        guard let regex = try? NSRegularExpression(pattern: self.pattern(), options: .caseInsensitive) else {
            print("Error generating block property finder regex. Regex: " + self.pattern())
            return nil
        }
        
        return regex
    }
    
    func makeContentHugoFriendly(content: String) -> String {
        guard !content.isEmpty, let regex = regex() else { return content }
        
        return regex.stringByReplacingMatches(in: content, range: content.range(), withTemplate: "")
    }
    
}


enum Shortcodes {
    case youtube
    case twitter
    case vimeo
    
    static func shortcodes() -> [Shortcodes] {
        return [
//            .youtube,
//            .twitter,
//            .vimeo
        ]
    }
    
    private func pattern() -> String {
        switch self {
        case .youtube:
            return #"\{\{youtube\s*(.*)\s*\}\}"#
        case .twitter:
            return #"\{\{twitter\s*(.*)\s*\}\}"#
        case .vimeo:
            return #"\{\{vimeo\s*(.*)\s*\}\}"#
        }
    }
    
    
    //TODO: this data extraction could be less hacky for all these
    private func processLink(_ link: String) -> String {
        switch self {
        case .youtube:
            return String(link.split(separator: "/").last ?? "\(link)")
        case .twitter:
            //converting "https://twitter.com/SanDiegoZoo/status/1453110110599868418"
            // to {{< tweet user="SanDiegoZoo" id="1453110110599868418" >}}
            
            var sections = link.split(separator: "/")
            let id = sections.popLast()
            _ = sections.popLast()
            let user = sections.popLast()
            
            if let id = id, let user = user {
                return "user=\"\(user)\" id=\"\(id)\""
            } else {
                return link
            }
        case .vimeo:
            return String(link.split(separator: "/").last ?? "\(link)")
        }
    }
    
    
    private func replacement() -> String {
        switch self {
        case .youtube:
            return #"{{< youtube $1 >}}"#
        case .twitter:
            return #"{{< tweet $1 >}}"#
        case .vimeo:
            return #"{{< vimeo $1 >}}"#
        }
    }
    
    private func regex() -> NSRegularExpression? {
        guard let regex = try? NSRegularExpression(pattern: self.pattern(), options: .caseInsensitive) else {
            print("Error generating shortcode finder regex. Regex: " + self.pattern())
            return nil
        }
        
        return regex
    }
    
    func makeContentHugoFriendly(content: String) -> String {
        guard !content.isEmpty, let regex = regex() else { return content }
        
        return regex.stringByReplacingMatches(in: content, range: content.range(), withTemplate: self.replacement())
    }
}


enum AssetFinder {
    case assetWithProperties
    case asset
    
    static func assetUpdates() -> [AssetFinder] {
        return [
            .assetWithProperties,
            .asset
        ]
    }
    
    static func assetsFolderName() -> String {
        return "assets"
    }
    
    func pattern() -> String {
        switch self {
        case .assetWithProperties:
            return AssetFinder.asset.pattern() + #"\{.*\}"#
        case .asset:
            return #"\(\.\.\/"# + AssetFinder.assetsFolderName() + #"\/(.*)\)"#
        }
    }
    
    func replacement() -> String {
        return #"\(/"# + AssetFinder.assetsFolderName() + #"/$1\)"#
    }
    
    private func regex() -> NSRegularExpression? {
        guard let regex = try? NSRegularExpression(pattern: self.pattern(), options: .caseInsensitive) else {
            print("Error generating asset finder regex, might be a bad asset folder name. Regex: " + self.pattern())
            return nil
        }
        
        return regex
    }
    
    func makeContentHugoFriendly(content: String) -> String {
        guard let regex = regex() else { return content }
        
        let replaced = regex.stringByReplacingMatches(in: content, options: .reportCompletion, range: content.range(), withTemplate: replacement())
        return replaced
    }
    
    static func extractAssetNames(fromContent content: String) -> [String] {
        guard !content.isEmpty, let regex = AssetFinder.asset.regex() else { return [] }
        
        return regex.matches(in: content, range: content.range()).compactMap { match in
            let assetNameRange = match.range(at: 1) //this is the first capture group
            if let substringRange = Range(assetNameRange, in: content) {
                return String(content[substringRange])
            } else{
                return nil
            }
        }
        
    }
}



enum LinkFinder {
    
    case pageEmbed(name: String, path: String)
    case pageAlias(name: String, path: String)
    case pageReference(name: String, path: String)
    
    case blockEmbed(uuid: String, content: String, path: String)
    case blockAlias(uuid: String, content: String, path: String)
    case blockReference(uuid: String, content: String, path: String)
    
    static func pageLinkChecks(name: String, path: String) -> [LinkFinder] {
        //order matters
        return [
            LinkFinder.pageEmbed(name: name, path: path),
            LinkFinder.pageAlias(name: name, path: path),
            LinkFinder.pageReference(name: name, path: path)
        ]
    }
    
    static func blockLinkChecks(uuid: String, content: String, path: String) -> [LinkFinder] {
        //order matters
        return [
            LinkFinder.blockEmbed(uuid: uuid, content: content, path: path),
            LinkFinder.blockAlias(uuid: uuid, content: content, path: path),
            LinkFinder.blockReference(uuid: uuid, content: content, path: path)
        ]
    }
    
    func pattern() -> String {
        switch self {
        case .pageEmbed(let name, _):
            return #"\{\{embed\s*\[\[\s*"# + name.escapeParentheses() + #"\s*\]\]\s*\}\}"#
        case .pageAlias(let name, _):
            return #"\]\(\s*\[\[\s*"# + name.escapeParentheses() + #"\s*\]\]s*\)"#
        case .pageReference(let name, _):
            return #"\[\[\s*"# + name.escapeParentheses() + #"\s*\]\]"#
            
        case .blockEmbed(let uuid, _, _):
            return #"\{\{embed\s*\(\(\s*"# + uuid + #"\s*\)\)\s*\}\}"#
        case .blockAlias(let uuid, _, _):
            return #"\]\(\s*\(\(\s*"# + uuid + #"\s*\)\)\s*\)"#
        case .blockReference(let uuid, _, _):
            return #"\(\(\s*"# + uuid + #"\s*\)\)"#
        }
    }
    
    func readable() -> String { //TODO: add a version that adds plain links
        switch self {
        case .pageEmbed(let name, _):
            return "[[\(name)]]"
        case .pageAlias:
            return "]"
        case .pageReference(let name, _):
            return "[[\(name)]]"
            
            
        case .blockEmbed(_, let content, _):
            return content
        case .blockAlias:
            return "]"
        case .blockReference(_, let content, _):
            return shortenedBlockContent(content: content)
        }
    }
    
    private func shortenedBlockContent(content: String) -> String {
        if let firstLine = content.split(separator: "\n").first {
            return String(firstLine)
        } else {
            return content
        }
    }
    
    func hugoFriendlyLink() -> String {
        switch self {
        case .pageEmbed(_, let path):
            return "{{< links/page-embed \"\(path)\" >}}"
        case .pageAlias(_, let path):
            return "](\(path))"
        case .pageReference(let name, let path):
            return "[\(name)](\(path))"
            
            
        case .blockEmbed(_, _, let path):
            return "{{< links/block-embed \"\(path)\" >}}"
        case .blockAlias(_, _, let path):
            return "](\(path))"
        case .blockReference(_, let content, let path):
            return "[\(shortenedBlockContent(content: content))](\(path))"
        }
    }
    
    func ranges(inContent content: String) -> [NSRange] {
        guard let regex = try? NSRegularExpression(pattern: self.pattern(), options: .caseInsensitive) else {
            print("Skipping finding links for bad regex: " + self.pattern())
            return []
        }
        return regex.matches(in: content, options: .reportCompletion, range: content.range()).map { (result: NSTextCheckingResult) in
            return result.range
        }
    }
    
    func makeContentHugoFriendly(_ content: String, noLinks: Bool) -> String {
        var updatedContent = content as NSString
        
        ranges(inContent: content).forEach { range in
            let replacement = noLinks ? readable() : hugoFriendlyLink()
            updatedContent = updatedContent.replacingCharacters(in: range, with: replacement) as NSString
        }
        
        return updatedContent as String
    }
}

extension String {
    func escapeParentheses() -> String {
        return replacingOccurrences(of: "(", with: #"\("#).replacingOccurrences(of: ")", with: #"\)"#)
    }
    
    func range() -> NSRange {
        return NSRange(startIndex..<endIndex, in: self)
    }
}

File: logseq_compiler/link_finder.py
import re
from typing import Optional, List

class BlockPropertyFinder:
    @staticmethod
    def make_content_hugo_friendly(content: str) -> str:
        if not content:
            return content
        pattern = r"\n\S+::\s+\S+"
        return re.sub(pattern, '', content, flags=re.IGNORECASE)

class Shortcodes:
    YOUTUBE = 'youtube'
    TWITTER = 'twitter'
    VIMEO = 'vimeo'

    @staticmethod
    def shortcodes() -> List[str]:
        return [Shortcodes.YOUTUBE, Shortcodes.TWITTER, Shortcodes.VIMEO]

    @staticmethod
    def pattern(shortcode: str) -> str:
        patterns = {
            Shortcodes.YOUTUBE: r"\{\{youtube\s*(.*?)\s*\}\}",
            Shortcodes.TWITTER: r"\{\{twitter\s*(.*?)\s*\}\}",
            Shortcodes.VIMEO: r"\{\{vimeo\s*(.*?)\s*\}\}",
        }
        return patterns[shortcode]

    @staticmethod
    def process_link(shortcode: str, link: str) -> str:
        if shortcode in [Shortcodes.YOUTUBE, Shortcodes.VIMEO]:
            return link.strip().split('/')[-1]
        elif shortcode == Shortcodes.TWITTER:
            # e.g. https://twitter.com/SanDiegoZoo/status/1453110110599868418
            parts = link.strip().split('/')
            if len(parts) >= 2:
                id_ = parts[-1]
                user = parts[-3] if len(parts) >= 3 else ''
                return f'user="{user}" id="{id_}"'
            return link
        else:
            return link

    @staticmethod
    def replacement(shortcode: str, inside: str) -> str:
        replacements = {
            Shortcodes.YOUTUBE: f"{{{{< youtube {inside} >}}}}",
            Shortcodes.TWITTER: f"{{{{< tweet {inside} >}}}}",
            Shortcodes.VIMEO: f"{{{{< vimeo {inside} >}}}}",
        }
        return replacements[shortcode]


class LinkFinder:
    PAGE_EMBED = 'page_embed'
    PAGE_ALIAS = 'page_alias'
    PAGE_REFERENCE = 'page_reference'
    BLOCK_EMBED = 'block_embed'
    BLOCK_ALIAS = 'block_alias'
    BLOCK_REFERENCE = 'block_reference'

    def __init__(self, kind: str, name: Optional[str] = None, uuid: Optional[str] = None, content: Optional[str] = None, path: Optional[str] = None):
        self.kind = kind
        self.name = name
        self.uuid = uuid
        self.content = content
        self.path = path

    @staticmethod
    def page_link_checks(name: str, path: str) -> List['LinkFinder']:
        # Order matters
        return [
            LinkFinder(LinkFinder.PAGE_EMBED, name=name, path=path),
            LinkFinder(LinkFinder.PAGE_ALIAS, name=name, path=path),
            LinkFinder(LinkFinder.PAGE_REFERENCE, name=name, path=path),
        ]

    @staticmethod
    def block_link_checks(uuid: str, content: str, path: str) -> List['LinkFinder']:
        # Order matters
        return [
            LinkFinder(LinkFinder.BLOCK_EMBED, uuid=uuid, content=content, path=path),
            LinkFinder(LinkFinder.BLOCK_ALIAS, uuid=uuid, content=content, path=path),
            LinkFinder(LinkFinder.BLOCK_REFERENCE, uuid=uuid, content=content, path=path),
        ]

    def pattern(self) -> str:
        if self.kind == self.PAGE_EMBED:
            return r"\{\{embed\s*\[\[\s*" + re.escape(self.name or '') + r"\s*\]\]\s*\}\}"
        elif self.kind == self.PAGE_ALIAS:
            # Aliased page: [alias]([[actual page]])
            return r"\]\(\s*\[\[\s*" + re.escape(self.name or '') + r"\s*\]\]\s*\)"
        elif self.kind == self.PAGE_REFERENCE:
            return r"\[\[\s*" + re.escape(self.name or '') + r"\s*\]\]"
        elif self.kind == self.BLOCK_EMBED:
            return r"\{\{embed\s*\(\(\s*" + re.escape(self.uuid or '') + r"\s*\)\)\s*\}\}"
        elif self.kind == self.BLOCK_ALIAS:
            # Aliased block: [alias](((uuid)))
            return r"\]\(\s*\(\(\s*" + re.escape(self.uuid or '') + r"\s*\)\)\s*\)"
        elif self.kind == self.BLOCK_REFERENCE:
            return r"\(\(\s*" + re.escape(self.uuid or '') + r"\s*\)\)"
        else:
            return ''

    def readable(self) -> str:
        if self.kind == self.PAGE_EMBED or self.kind == self.PAGE_REFERENCE:
            return f"[[{self.name}]]"
        elif self.kind == self.PAGE_ALIAS:
            return "]"
        elif self.kind == self.BLOCK_EMBED:
            return self.content or ''
        elif self.kind == self.BLOCK_ALIAS:
            return "]"
        elif self.kind == self.BLOCK_REFERENCE:
            return self.shortened_block_content(self.content or '')
        else:
            return ''

    def shortened_block_content(self, content: str) -> str:
        return content.split('\n', 1)[0] if '\n' in content else content

    def hugo_friendly_link(self) -> str:
        if self.kind == self.PAGE_EMBED:
            return f"{{{{< links/page-embed \"{self.path}\" >}}}}"
        elif self.kind == self.PAGE_ALIAS:
            return f"]({self.path})"
        elif self.kind == self.PAGE_REFERENCE:
            return f"[{self.name}]({self.path})"
        elif self.kind == self.BLOCK_EMBED:
            return f"{{{{< links/block-embed \"{self.path}\" >}}}}"
        elif self.kind == self.BLOCK_ALIAS:
            return f"]({self.path})"
        elif self.kind == self.BLOCK_REFERENCE:
            return f"[{self.shortened_block_content(self.content or '')}]({self.path})"
        else:
            return ''

    def ranges(self, content: str) -> List[tuple]:
        pattern = self.pattern()
        matches = list(re.finditer(pattern, content, flags=re.IGNORECASE))
        return [(m.start(), m.end()) for m in matches]

    def make_content_hugo_friendly(self, content: str, no_links: bool = False) -> str:
        updated_content = content
        for _ in range(len(self.ranges(updated_content))):
            matches = list(re.finditer(self.pattern(), updated_content, flags=re.IGNORECASE))
            if matches:
                match = matches[0]
                replacement = self.readable() if no_links else self.hugo_friendly_link()
                updated_content = (
                    updated_content[:match.start()] + replacement + updated_content[match.end():]
                )
        return updated_content
